pssh/ssh_client.py

```diff
@@ -346,6 +342,77 @@ def copy_file(self, local_file, remote_file, recurse=False):
         except Exception, error:
             logger.error("Error occured copying file %s to remote destination %s:%s - %s",
                          local_file, self.host, remote_file, error)
+            raise error
         else:
             logger.info("Copied local file %s to remote destination %s:%s",
                         local_file, self.host, remote_file)
+
+    def _copy_dir_to_local(self, remote_dir, local_dir):
+        """Copies the remote directory to the local host."""
+        sftp = self._make_sftp()
+        file_list = sftp.listdir(remote_dir)
+        for file_name in file_list:
+            remote_path = os.path.join(remote_dir, file_name)
+            local_path = os.path.join(local_dir, file_name)
+            self.copy_file_to_local(remote_path, local_path, recurse=True)
+
+    def copy_file_to_local(self, remote_file, local_file, recurse=False):
+        """Copy remote file to local host via SFTP/SCP
+
+        Copy is done natively using SFTP/SCP version 2 protocol, no scp command \
+        is used or required.
+
+        :param remote_file: Remote filepath to copy the file from.
+        :type remote_file: str
+        :param local_file: Local filepath where the file will be copied.
+        :type local_file: str
+        :param recurse: Whether or not to recursively copy directories.
+        :type recurse: bool
+
+        :raises: :mod:`ValueError` when a directory is supplied to remote_file \
+        and recurse is not set
+        :raises: :mod:`OSError` on OS errors creating directories or file
+        :raises: :mod:`IOError` on IO errors creating directories or file
+        """
+        sftp = self._make_sftp()
+        try:
+            sftp.listdir(remote_file)
+        except (OSError, IOError):
+            remote_dir_exists = False
+        else:
+            remote_dir_exists = True
+        if remote_dir_exists and recurse:
+            return self._copy_dir_to_local(remote_file, local_file)
+        elif remote_dir_exists and not recurse:
+            raise ValueError("Recurse must be true if remote_file is a "
+                             "directory.")
+        destination = self._parent_path_split(local_file)
+        self._make_local_dir(destination)
```

Comment:
No worries, take your time, took me a while to get around to the PR :) 

FYI, there is a recursive sftp `mkdir` in `SSHClient` that can be used as a template for the same with `os.mkdir`.